File: github_code/minimum-spanning-tree/java/65.java
package io.acoia.graphs;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import com.google.common.graph.EndpointPair;
import com.google.common.graph.MutableValueGraph;
import com.google.common.graph.ValueGraph;
import com.google.common.graph.ValueGraphBuilder;

import io.acoia.sets.UnionFind;

/**
 * Algorithms for computing the minimum spanning tree of graphs.
 * 
 * Copyright (c) 2017 Royce Ausburn This code is available for use in accordance with the MIT
 * License https://opensource.org/licenses/MIT
 * 
 * @author Royce Ausburn (esapersona@royce.id.au)
 * 
 */
public class MinimumSpanningTree {

  /**
   * Sorts the edges of the provided ValueGraph and then calls kruskals(ValueGraph<N, E> g,
   * List<EndpointPair<N>> sortedEdges). Due to the need to sort the edges this method requires edge
   * values in the graph to implement Comparable.
   * 
   * Due to the need to sort this method sorts the edges (O(E log(E))) before calling Kruskal's (O(E α(V))) .
   */
  public static <N, E extends Comparable<E>> ValueGraph<N, E> kruskals(ValueGraph<N, E> g) {
    Set<EndpointPair<N>> edges = g.edges();

    List<EndpointPair<N>> sortedEdges =
        edges.stream()
          .sorted((ep1, ep2) -> g.edgeValue(ep1.nodeU(), ep1.nodeV()).compareTo(g.edgeValue(ep2.nodeU(), ep2.nodeV())))
          .collect(Collectors.toList());

    return kruskals(g, sortedEdges);
  }

  /**
   * An implementation of Kruskal's minimum spanning tree algorithm.
   * 
   * Description from Wiki: Kruskal's algorithm is a minimum-spanning-tree algorithm which finds an
   * edge of the least possible weight that connects any two trees in the forest. It is a greedy
   * algorithm in graph theory as it finds a minimum spanning tree for a connected weighted graph
   * adding increasing cost arcs at each step. This means it finds a subset of the edges that forms
   * a tree that includes every vertex, where the total weight of all the edges in the tree is
   * minimized. If the graph is not connected, then it finds a minimum spanning forest (a minimum
   * spanning tree for each connected component).
   * 
   * This implementation uses io.acoia.sets.UnionFind and expects a sorted list of edges, and thus
   * runs in O(E α(V)) time.
   */
  public static <E, N> ValueGraph<N, E> kruskals(ValueGraph<N, E> g, List<EndpointPair<N>> sortedEdges) {

    if (g.isDirected())
      throw new IllegalArgumentException("Kruskal's does not support directed graphs");

    // Set up a Graph to store the result. 
    MutableValueGraph<N, E> tree = ValueGraphBuilder
        .undirected()
        .allowsSelfLoops(false)
        .expectedNodeCount(g.nodes().size())
        .build();

    for (N n : g.nodes()) {
      tree.addNode(n);
    }

    // Now we start kruskals.. most of the work is done by the UnionFind really...
    UnionFind<N> uf = new UnionFind<>(g.nodes());
    for (EndpointPair<N> e : sortedEdges) {
      
      N nodeU = e.nodeU();
      N nodeV = e.nodeV();
      
      if (uf.findRoot(nodeU) != uf.findRoot(nodeV)) {
        tree.putEdgeValue(nodeU, nodeV, g.edgeValue(nodeU, nodeV));
        uf.join(nodeU, nodeV);
      }
    }

    return tree;
  }

}

File: github_code/minimum-spanning-tree/cpp/65.c
/*
Kruskal's MST Algorithm
-----------------------
Kruskal's algorithm is a minimum-spanning-tree algorithm which
 finds an edge of the least possible weight that connects any 
 two trees in the forest.[1] It is a greedy algorithm in graph 
 theory as it finds a minimum spanning tree for a connected 
 weighted graph adding increasing cost arcs at each step.
-----------------------
Time Complexity : O(m*logn)
Space Complexity : O(n)
*/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int n, m;

struct edge{
	int start, end;
	int weight;
	bool operator <(const edge &e) const{
		return (weight < e.weight);
	}
};

struct edge edges[100];

struct node{
	int parent;
	int rank;
};

struct node disjoint_set_union[100];

void makeset(int x){
	disjoint_set_union[x].parent = x;
	disjoint_set_union[x].rank = 0;
}

int find(int x){
	if (disjoint_set_union[x].parent == x){
		return x;
	}
	disjoint_set_union[x].parent = find(disjoint_set_union[x].parent);
	return disjoint_set_union[x].parent;
}

void join(int x, int y){
	int x_root = find(x);
	int y_root = find(y);
	if (x_root == y_root){
		return ;
	}
	if (disjoint_set_union[x_root].rank < disjoint_set_union[y_root].rank){
		disjoint_set_union[x_root].parent = y_root;
	} else if (disjoint_set_union[x_root].rank > disjoint_set_union[y_root].rank){
		disjoint_set_union[y_root].parent = x_root;
	} else{
		disjoint_set_union[y_root].parent = x_root;
		disjoint_set_union[x_root].rank++;
	}
}

int kruskal(){
	int returned = 0, num_components = n;
	for (int i = 0; i < n; i++){
		makeset(i);
	}
	sort(edges, edges + m);
	for (int i = 0; i < m and num_components > 1; i++){
		if (find(edges[i].start) != find(edges[i].end)){
			join(edges[i].start, edges[i].end);
			returned += edges[i].weight;
			num_components--; 
		}
	}
	if (num_components > 1){
		return -1;
	}
	return returned;
}
